### PR TITLE
[AF-2339] Add support for branch management in the rest API

### PR DESCRIPTION
**Pull request by gabriel376**
[AF-2339 Add support for branch management in the rest API](https://issues.jboss.org/browse/AF-2339)

Added 3 new endpoints to REST API:
1. **GET** /spaces/{spaceName}/projects/{projectName}/branches
Get a list of branches for **projectName** on **spaceName**
2. **POST** /spaces/{spaceName}/projects/{projectName}/branches
Add a new branch for **projectName** on **spaceName**
3. **DELETE** /spaces/{spaceName}/projects/{projectName}/branches/{branchName}
Remove **branchName** for **projectName** on **spaceName**

Methods **addBranch()** and **removeBranch()** on **WorkspaceProjectServiceImpl.java** were migrated from **kie-wb-common/LibraryServiceImpl.java**

* Depends on https://github.com/kiegroup/kie-wb-common/pull/3049

### CODE REVIEW

File: uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java [hunk 1]
```diff
@@ -498,6 +499,74 @@ public JobResult createSpace(final String jobId,
...
+        }
+
+        try {
+            workspaceProjectService.addBranch(newBranchName,
+                                              baseBranchName,
+                                              project);
+            result.setStatus(JobStatus.SUCCESS);
+
+        } catch (FileAlreadyExistsException e) {
+            result.setStatus(JobStatus.BAD_REQUEST);
```
Comment (paulovmr) [inline uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java]:
```suggestion
            result.setStatus(JobStatus.DUPLICATE_RESOURCE);
```

File: uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java [hunk 2]
```diff
@@ -498,6 +499,74 @@ public JobResult createSpace(final String jobId,
...
+            result.setResult("Branch [" + newBranchName + "] already exists.");
+
+        } catch (Exception e) {
+            result.setStatus(JobStatus.BAD_REQUEST);
```
Comment (paulovmr) [inline uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java]:
```suggestion
            result.setStatus(JobStatus.FAIL);
```

File: uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java [hunk 3]
```diff
@@ -498,6 +499,74 @@ public JobResult createSpace(final String jobId,
...
+            return result;
+        }
+
+        try {
+            workspaceProjectService.removeBranch(branchName,
+                                                 project);
+            result.setStatus(JobStatus.SUCCESS);
+
+        } catch (Exception e) {
+            result.setStatus(JobStatus.BAD_REQUEST);
```
Comment (paulovmr) [inline uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java]:
```suggestion
            result.setStatus(JobStatus.FAIL);
```

File: uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java [hunk 4]
```diff
@@ -171,6 +172,83 @@ public void testsFail() throws Exception {
...
+    public void testAddBranchBadRequest() {
+        doThrow(Exception.class).when(workspaceProjectService).addBranch(any(), any(), any());
+
+        JobResult jobResult = helper.addBranch(null,
+                                               space.getName(),
+                                               "project",
+                                               "new-branch",
+                                               "ref-branch");
+
+        assertEquals(JobStatus.BAD_REQUEST,
```
Comment (paulovmr) [inline uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java]:
```suggestion
        assertEquals(JobStatus.FAIL,
```

File: uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java [hunk 5]
```diff
@@ -171,6 +172,83 @@ public void testsFail() throws Exception {
...
+    @Test
+    public void testRemoveBranchBadRequest() {
+        doThrow(Exception.class).when(workspaceProjectService).removeBranch(any(), any());
+
+        JobResult jobResult = helper.removeBranch(null,
+                                               space.getName(),
+                                               "project",
+                                               "new-branch");
+
+        assertEquals(JobStatus.BAD_REQUEST,
```
Comment (paulovmr) [inline uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java]:
```suggestion
        assertEquals(JobStatus.FAIL,
```

File: uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java (hunk 1 above)
Comment (gabriel376) [reply uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java]: Done

File: uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java (hunk 2 above)
Comment (gabriel376) [reply uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java]: Done

File: uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java (hunk 4 above)
Comment (gabriel376) [reply uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java]: Done

File: uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java (hunk 5 above)
Comment (gabriel376) [reply uberfire-rest/uberfire-rest-backend/src/test/java/org/guvnor/rest/backend/JobRequestHelperTest.java]: Done

File: uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java (hunk 3 above)
Comment (gabriel376) [reply uberfire-rest/uberfire-rest-backend/src/main/java/org/guvnor/rest/backend/JobRequestHelper.java]: Done

**Comment by paulovmr** [comment]
jenkins execute full downstream build

**Comment by barboras7** [comment]
jenkins execute full downstream build

**Comment by paulovmr** [comment]
Hi @barboras7 , I saw that you approved the other related PRs except for this one, do you have any observations?

**Comment by barboras7** [comment]
jenkins execute full downstream build

**Comment by barboras7** [comment]
@paulovmr, @gabriel376,
sorry guys, I was busy with 7.6.0 before Christmas. I checked also this PR and it looks good, just one question regarding the usage. I am getting "status: APPROVED" for operations where I would not expect it. Let me share an example:

`curl -u "testadmin:admin1234;" -H "Content-Type: application/json" -X POST -d '{"newBranchName":"testBranch1", "baseBranchName":"master"}' http://localhost:8080/business-central/rest/spaces/testspace/projects/testproject/branches`

`{"jobId":"1578040007622-8","status":"APPROVED","spaceName":"testspace","projectName":"testproject","newBranchName":"testBranch1","baseBranchName":"master","userIdentifier":"testadmin"}`

This is what I get when I create the "testBranch1" for the second time (I run the POST command twice) which means that the testBranch1 was already present and therefore not created the second time. Is "APPROVED" expected behavior?

**Comment by gabriel376** [comment]
Hi @barboras7,

According to test case in kie-wb-distributions, the REST API returns DUPLICATE_RESOURCE when you try to add a branch twice
https://github.com/kiegroup/kie-wb-distributions/pull/999/files#diff-d626ee084dbca1fa18ca6cba8c3811efR337
What WAR file are you using ?

**Comment by barboras7** [comment]
@gabriel376,
yes, I know, that's why I am asking. I am using the war file from full downstream build. I tried once again and I got the following responses:

CREATE THE 'testBranch' FOR THE FIRST TIME:
`curl -u "testadmin:admin1234;" -H "Content-Type: application/json" -X POST -d '{"newBranchName":"testBranch", "baseBranchName":"master"}' http://localhost:8080/business-central/rest/spaces/testspace/projects/testproject/branches`

RESPONSE FOR FIRST CREATE:
`{"jobId":"1578061404672-1","status":"APPROVED","spaceName":"testspace","projectName":"testproject","newBranchName":"testBranch","baseBranchName":"master","userIdentifier":"testadmin"}`

LIST ALL BRANCHES:
`curl -u "testadmin:admin1234;" -H "Content-Type: application/json" -X GET http://localhost:8080/business-central/rest/spaces/testspace/projects/testproject/branches
[{"name":"testBranch"},{"name":"master"}]`

CREATE THE 'testBranch' FOR THE SECOND TIME:
`curl -u "testadmin:admin1234;" -H "Content-Type: application/json" -X POST -d '{"newBranchName":"testBranch", "baseBranchName":"master"}' http://localhost:8080/business-central/rest/spaces/testspace/projects/testproject/branches`

RESPONSE FOR SECOND CREATE:
`{"jobId":"1578061427597-2","status":"APPROVED","spaceName":"testspace","projectName":"testproject","newBranchName":"testBranch","baseBranchName":"master","userIdentifier":"testadmin"}`

If the second response is not correct, could you please test this as well?

**Comment by gabriel376** [comment]
Hi @barboras7,

I just tested again and it's working fine. You always get APPROVED when adding a branch because it returns a Job, so it means the Job was approved and will be processed. You should see Job status to check if Job was successful or not:

```
curl -X GET  -u '<USER>:<PASS>' "<URL>/rest/jobs/<JOB_ID>"
```

Possible responses are:
```
{"status":"SUCCESS","jobId":"1578070219736-3","result":null,"lastModified":1578070219775,"detailedResult":null}

{"status":"DUPLICATE_RESOURCE","jobId":"1578070233520-4","result":"Branch [b1] already exists.","lastModified":1578070233539,"detailedResult":null}
```

**Comment by paulovmr** [comment]
jenkins execute full downstream build

**Comment by gitgabrio** [comment]
@gabriel376 @ederign @manstis @danielezonca @yesamer 
Hi all, just found this PR has already been merged, anyway.
I see that this PR adds two methods inside an interface, then those methods are fully implemented only in one implementation, but in another one (the "mocked" one inside kogito common webapp) they have been provided empty, to comply with interface declaration.
IMHO this is not the best possible approach, for two reasons
1) it lead to pollute code with useless/empty methods
2) it breaks all possible descendant that are outside our visibility (being all that open source code, I should think that someone else in the world may have implemented the same interface)

In such situations I think we should provide "default" method implementation inside the interfaces. This would provide backward compatibility (i.e. it does not break someone else' code) and avoid creating empty/unneeded methods all over the codebase. 

MTC.
Best
Gabriele

**Pull request merged**